File: Stack/1. Valid Parentheses/solution.scala
object Solution {
  def isValid(s: String): Boolean = {
    import scala.collection.mutable
    val pairs = Map(')' -> '(', '}' -> '{', ']' -> '[')
    val stack = new mutable.Stack[Char](s.length)

    s.foreach {
      case c if pairs.contains(c) =>
        if (stack.isEmpty || pairs(c) != stack.pop()) {
          return false
        }
      case c => stack.push(c)
    }

    stack.isEmpty
  }
}

File: Stack/1. Valid Parentheses/solution.exs
defmodule Solution do
  @spec is_valid(s :: String.t) :: boolean
  def is_valid(s) do
    s
    |> String.graphemes()
    |> stack([])
  end

  def stack([], []), do: true

  def stack([")" | tail], ["(" | stack]) do
    stack(tail, stack)
  end

    def stack(["]" | tail], ["[" | stack]) do
    stack(tail, stack)
  end

    def stack(["}" | tail], ["{" | stack]) do
    stack(tail, stack)
  end

  def stack([h | tail], stack) when h in ["(", "{", "["] do
    stack(tail, [h | stack])
  end

  def stack(_, _), do: false
end
